cpp/src/strings/regex/regcomp.cpp

```diff
@@ -570,182 +570,180 @@ class regex_parser {
+          _prog.inst_at(id_inst1).u1.right_id  = id_inst2;
+          push_and(op.id_first, id_inst2);
           break;
-        case QUEST:
-          op2                                    = popand('?');
-          id_inst1                               = m_prog.add_inst(OR);
-          id_inst2                               = m_prog.add_inst(NOP);
-          m_prog.inst_at(id_inst1).u2.left_id    = id_inst2;
-          m_prog.inst_at(id_inst1).u1.right_id   = op2.id_first;
-          m_prog.inst_at(op2.id_last).u2.next_id = id_inst2;
-          pushand(id_inst1, id_inst2);
+        }
+        case QUEST: {
+          auto const op                        = pop_and();
+          auto const id_inst1                  = _prog.add_inst(OR);
+          auto const id_inst2                  = _prog.add_inst(NOP);
+          _prog.inst_at(id_inst1).u2.left_id   = id_inst2;
+          _prog.inst_at(id_inst1).u1.right_id  = op.id_first;
+          _prog.inst_at(op.id_last).u2.next_id = id_inst2;
+          push_and(id_inst1, id_inst2);
           break;
-        case QUEST_LAZY:
-          op2                                    = popand('?');
-          id_inst1                               = m_prog.add_inst(OR);
-          id_inst2                               = m_prog.add_inst(NOP);
-          m_prog.inst_at(id_inst1).u2.left_id    = op2.id_first;
-          m_prog.inst_at(id_inst1).u1.right_id   = id_inst2;
-          m_prog.inst_at(op2.id_last).u2.next_id = id_inst2;
-          pushand(id_inst1, id_inst2);
+        }
+        case QUEST_LAZY: {
+          auto const op                        = pop_and();
+          auto const id_inst1                  = _prog.add_inst(OR);
+          auto const id_inst2                  = _prog.add_inst(NOP);
+          _prog.inst_at(id_inst1).u2.left_id   = op.id_first;
+          _prog.inst_at(id_inst1).u1.right_id  = id_inst2;
+          _prog.inst_at(op.id_last).u2.next_id = id_inst2;
+          push_and(id_inst1, id_inst2);
           break;
+        }
       }
     }
   }
 
-  void Operator(int t)
+  void handle_operator(int token, int subid = 0)
   {
-    if (t == RBRA && --nbra < 0)
+    if (token == RBRA && --_bracket_count < 0) {
       // unmatched right paren
       return;
-    if (t == LBRA) {
-      nbra++;
-      if (lastwasand) Operator(CAT);
-    } else
-      evaluntil(t);
-    if (t != RBRA) pushator(t);
-    lastwasand = (t == STAR || t == QUEST || t == PLUS || t == STAR_LAZY || t == QUEST_LAZY ||
-                  t == PLUS_LAZY || t == RBRA);
+    }
+    if (token == LBRA) {
+      _bracket_count++;
+      if (_last_was_and) { handle_operator(CAT, subid); }
+    } else {
+      eval_until(token);
+    }
+    if (token != RBRA) { push_operator(token, subid); }
+
+    static std::vector<int> tokens{STAR, STAR_LAZY, QUEST, QUEST_LAZY, PLUS, PLUS_LAZY, RBRA};
+    _last_was_and =
+      std::any_of(tokens.begin(), tokens.end(), [token](auto t) { return t == token; });
```

Comment:
I guess tokens are cheap (and this is host code) so this doesn't really matter, but you could capture the token by reference to avoid copying right? Also, you could iterator using `cbegin` and `cend` instead and make the function parameter `auto const t`.

Reply:
Copy by value for int types is preferred over by reference if the value is not changed. I think the `cbegin/cend` is a good idea. I don't think declaring `t` as `const` is that helpful here. The compiler will generate the same code and a human only has one line of code to parse.

Reply:
I'm dumb, when I wrote "I guess tokens are cheap" I didn't check the vector above to see exactly what tokens are. Yes, definitely agree that copy by value is preferred here. I also agree that the const is not necessary, was just a minor suggestion.

Should we make the `tokens` vector constexpr, then? Also, just a note that if we ever switch `OperatorType` to be a C++-style scoped enum we should probably add a base type specifier that aligns with the type of the tokens vector just to be explicit and avoid type conversions.

Reply:
I think `constexpr` would be more appropriate. Unfortunately you can only declare `constexpr` for literal or reference types. 
So I prefer using `static` here.
It looks like `c++20` will support `constexpr` with `std::vector`.

Reply:
Right, that's a good point. This would probably be a good place to use `std::array` (with CTAD so that you don't even need to specify the length).